### PR TITLE
Not split nodes when searching for nodes but doing it all at once

### PR DESCRIPTION
**Pull request by wgliang**
**What this PR does / why we need it**:
Not split nodes when searching for nodes but doing it all at once.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
@bsalamat 

This is a follow up PR of #66733.

https://github.com/kubernetes/kubernetes/pull/66733#discussion_r205932531

**Release note**:

```release-note
Not split nodes when searching for nodes but doing it all at once.
```

### CODE REVIEW

**Comment by wgliang** [comment]
/assign @bsalamat 
PTAL

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go
Comment (misterikkit) [inline staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: The normal way to accomplish something like this in golang would be to use a `context.Context` for the work that you are dispatching, then cancelling the context when you want to abandon the rest of the work.

It would also make sense to have the workers fan in their results to a goroutine that handles results and decides when to cancel.

**Review by misterikkit — COMMENTED** [review]
I don't think we should try too hard to mimic the `Parallelize` API that is being replaced here. If there's a pattern that works better for us here, let's go with it.

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go
Comment (bsalamat) [inline staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: Given that this is in client-go, I think we should make this a more generic function. More specifically, I think we should:
1. Rename it to `ParallelizeUntil`.
1. Change `feasible *int32` to a function that returns a bool and an error, similar to [ConditionFunc](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go#L172). When the return value is true, it stops.

Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: Thanks for your kind review. :). I've updated it.

Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: Indeed, I understand what you mean, maybe a callback function can do the same, just like bsalamat said. And using Context may be too complicated for a public function.

**Review by misterikkit — COMMENTED** [review]
> I don't think we should try too hard to mimic the Parallelize API that is being replaced here. If there's a pattern that works better for us here, let's go with it.

I apologize. I didn't notice that you were adding the util in `client-go`. Given that, it makes lots of sense to mimic the Parallelize API.

> Indeed, I understand what you mean, maybe a callback function can do the same, just like bsalamat said. And using Context may be too complicated for a public function.

I don't believe that Context should be considered "too complicated" for a public function. In this case, it offers features that we don't need, so a stop channel would also work. (Passing a stop channel around is a very common pattern in k8s, but it's an old pattern that really ought to be replaced with Context.)

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go
Comment (k82cn) [inline staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: what happen if err is NOT nil?

Comment (k82cn) [inline staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: btw, as an util, do we accept nil conditionFunc ?

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go
Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: A good idea.

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go
Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: No problem.

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go
Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: After this PR merge, I will submit a separate PR to replace all Parallelize.

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go
Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: Yes,we should handle it. I'll follow misterikkit's comment and update it later.

File: pkg/scheduler/core/generic_scheduler.go
Comment (wgliang) [reply pkg/scheduler/core/generic_scheduler.go]: I have a question and need your guidance. If pass the channel instead of synchronization as you said, is it also needed for errs and failedPredicateMap?

Comment (wgliang) [reply pkg/scheduler/core/generic_scheduler.go]: And the size of the channel buffer is also a problem that needs to be determined.

Comment (wgliang) [reply pkg/scheduler/core/generic_scheduler.go]: Yes, I agree with you. We should pass in a channel, and then each worker function only needs data from the goroutine (here is one *v1.Node). But do we need to build this channel first? Because this may be a channel with a large size, this is my doubt.

File: pkg/scheduler/core/generic_scheduler.go
Comment (bsalamat) [inline pkg/scheduler/core/generic_scheduler.go]: If we re-write this if statement as below, we can change line 377 to `filtered = make([]*v1.Node, numNodesToFind)`. This saves us some memory.

```go
if fits {
	len := atomic.AddInt32(&filteredLen, 1)
	if len > numNodesToFind {
		cancel()
	} else {
		filtered[len - 1] = g.cachedNodeInfoMap[nodeName].Node()		
	}
}
```

Comment (wgliang) [reply pkg/scheduler/core/generic_scheduler.go]: DONE.

File: pkg/scheduler/core/generic_scheduler.go
Comment (wgliang) [reply pkg/scheduler/core/generic_scheduler.go]: I'm sorry,this is what I tested and I forgot to delete it.

**Comment by wgliang** [comment]
/ping @bsalamat @k82cn
for approve

**Review by bsalamat — APPROVED** [review]
/lgtm

**Comment by bsalamat** [comment]
@kubernetes/client-go-maintainers for approval

**Comment by k82cn** [comment]
LGTM, thanks :)

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go
Comment (sttts) [inline staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: does it make sense to express this through the new one?

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go
Comment (sttts) [inline staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: isn't ctx always the first arg by Golang convention?

Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: Yes, indeed. I can change it as the first ctx.

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go
Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: What do you think it can be? I have no idea.

Comment (sttts) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: Isn't `Parallelize(...) = ParallelizeUntil(nil, ...)`?

Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: Emmm, you can think of it this way, there will be a separate PR that will replace all Parallelize with ParallelizeUntil.
just like misterikkit said:https://github.com/kubernetes/kubernetes/pull/67555#discussion_r212155466

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go
Comment (Huang-Wei) [inline staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: maybe change to `return Parallelize(workers, pieces, doWorkPiece)` so that we don't need to check if context is done in `select{}` for this "ctx == nil" case

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go
Comment (misterikkit) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: I think there was a slight misunderstanding. We should only replace the *definition* of `Parallelize` so that existing callers won't see any change. `Parallelize` would just call `ParallelizeUntil`.

I think that is a small change that should be included in this PR.

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go
Comment (misterikkit) [inline staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: Suggest rewording "until the context done or canceled" to "until done or the context is canceled"

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go
Comment (misterikkit) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: Per other comments, I think the plan is for `Parallelize` to call this func with a nil context to eliminate duplicate code.

`context.TODO()` implies that we need further code changes to get an appropriate context. Since your goal is to create a context that never cancels, it is more appropriate to use `context.Background()`.

However, I think it would be cleaner to write it as,
```golang
var stop chan struct{}
if ctx != nil {
    stop = ctx.Done()
}
```

File: pkg/scheduler/core/generic_scheduler.go
Comment (misterikkit) [inline pkg/scheduler/core/generic_scheduler.go]: This comment needs updating, since `remainingNodesNumber` is no longer used.

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go
Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: This is not a good idea, ParallelizeUntil should not rely on Parallelize.Because ParallelizeUntil may replace Parallelize.

**Review by misterikkit — CHANGES_REQUESTED** [review]
As a side note, it would be nice to rewrite the commit description and PR description to more clearly explain the change.

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go
Comment (Huang-Wei) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: That's fair.

File: pkg/scheduler/core/generic_scheduler.go
Comment (wgliang) [reply pkg/scheduler/core/generic_scheduler.go]: Done, thanks.

**Comment by bsalamat** [comment]
/assign @sttts 

for another round of review and approval

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go
Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: @sttts The purpose of ParallelizeUntil is to allow the caller to control through the context. This PR is already using ParallelizeUntil.
The existence of Parallelize is temporary, as other packages are still using Parallelize. After this PR, I will submit a PR delete Parallelize separately.
I don't know if you understand what I mean. :)

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go
Comment (sttts) [inline staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: nit: would leave these lines.

**Comment by sttts** [comment]
/lgtm
/approve

For client-go changes.

**Comment by wgliang** [comment]
/ping @bsalamat
for milestone v1.12

**Pull request merged**
